File: frontend/js/views/promissory_note.js
// Promissory note generator — amortizes a fixed loan and assembles the note's
// clauses, via /calc/promissory-note. Previews live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
let LAST_DOC = null;

export async function renderPromissoryNote(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.pnote.h1.title">// PROMISSORY NOTE</span></h1>
        <p class="muted small" data-i18n="view.pnote.hint.intro">
            A borrower's written promise to repay a fixed loan — for owner financing, intra-family
            loans, business notes, or seller carry-back. It amortizes the loan (level monthly payment,
            total interest, maturity date) and assembles the operative clauses: promise to pay, interest,
            repayment, prepayment, default/acceleration, and governing law. Drafting aid, not legal advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.pnote.h2.inputs">Note terms</h2>
            <form id="pnote-form" class="inline-form">
                <label><span data-i18n="view.pnote.label.state">Governing state</span>
                    <input type="text" name="governing_state" value="Texas" required></label>
                <label><span data-i18n="view.pnote.label.lender">Lender name</span>
                    <input type="text" name="lender_name" value=""></label>
                <label><span data-i18n="view.pnote.label.borrower">Borrower name</span>
                    <input type="text" name="borrower_name" value=""></label>
                <label><span data-i18n="view.pnote.label.principal">Principal ($)</span>
                    <input type="number" step="0.01" min="0" name="principal_usd" value="10000" required></label>
                <label><span data-i18n="view.pnote.label.rate">Annual rate (%)</span>
                    <input type="number" step="0.001" min="0" name="annual_rate_pct" value="6" required></label>
                <label><span data-i18n="view.pnote.label.term">Term (months)</span>
                    <input type="number" step="1" min="1" name="term_months" value="12" required></label>
                <label><span data-i18n="view.pnote.label.start">Note date</span>
                    <input type="date" name="start_date" value="2026-01-01" required></label>
                <label><span data-i18n="view.pnote.label.late_fee">Late fee per installment ($)</span>
                    <input type="number" step="0.01" min="0" name="late_fee_usd" value="0"></label>
                <label><span data-i18n="view.pnote.label.statute">Statute citation (optional)</span>
                    <input type="text" name="statute_citation" value="" placeholder="${esc(t('view.pnote.ph.statute'))}"></label>
            </form>
        </div>
        <div id="pnote-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#pnote-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            lender_name: (fd.get('lender_name') || '').trim(),
            borrower_name: (fd.get('borrower_name') || '').trim(),
            principal_usd: Number(fd.get('principal_usd')) || 0,
            annual_rate_pct: Number(fd.get('annual_rate_pct')) || 0,
            term_months: Math.round(Number(fd.get('term_months')) || 0),
            start_date: fd.get('start_date'),
            late_fee_usd: Number(fd.get('late_fee_usd')) || 0,
            governing_state: (fd.get('governing_state') || '').trim(),
            statute_citation: (fd.get('statute_citation') || '').trim(),
        };
        try {
            const doc = await api.calcPromissoryNote(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.pnote.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const el = mount.querySelector('#pnote-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.pnote.card.payment">Monthly payment</div>
                    <div class="value">${money(doc.monthly_payment_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.pnote.card.interest">Total interest</div>
                    <div class="value">${money(doc.total_interest_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.pnote.card.total">Total of payments</div>
                    <div class="value">${money(doc.total_of_payments_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.pnote.card.maturity">Maturity</div>
                    <div class="value">${esc(doc.maturity_date || '—')}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="pnote-copy" type="button" data-i18n="view.pnote.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="pnote-download" type="button" data-i18n="view.pnote.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#pnote-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.pnote.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.pnote.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#pnote-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'promissory-note.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/promissory_note.rs
//! Promissory note — a borrower's written promise to repay a fixed loan on
//! defined terms. Used for owner financing, intra-family loans, business notes,
//! and seller carry-back. This amortizes the loan (level monthly payment, total
//! interest, total of payments, maturity date) and assembles the note's
//! operative clauses. Drafting aid, not legal advice.

use chrono::{Months, NaiveDate};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct NoteInput {
    pub lender_name: String,
    pub borrower_name: String,
    pub principal_usd: f64,
    pub annual_rate_pct: f64,
    pub term_months: u32,
    /// Note date / first payment basis (YYYY-MM-DD).
    pub start_date: String,
    /// Flat late charge per missed installment (0 = none).
    #[serde(default)]
    pub late_fee_usd: f64,
    pub governing_state: String,
    #[serde(default)]
    pub statute_citation: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct PromissoryNote {
    pub title: String,
    pub monthly_payment_usd: f64,
    pub total_of_payments_usd: f64,
    pub total_interest_usd: f64,
    pub maturity_date: String,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

/// Level monthly payment on a fully-amortizing loan. Zero rate → straight
/// principal split; zero term → zero payment.
pub fn monthly_payment(principal: f64, annual_rate_pct: f64, term_months: u32) -> f64 {
    if term_months == 0 {
        return 0.0;
    }
    let n = term_months as i32;
    let r = annual_rate_pct / 100.0 / 12.0;
    if r == 0.0 {
        principal / n as f64
    } else {
        principal * r / (1.0 - (1.0 + r).powi(-n))
    }
}

pub fn generate(i: &NoteInput) -> PromissoryNote {
    let payment = monthly_payment(i.principal_usd, i.annual_rate_pct, i.term_months);
    let total_of_payments = payment * i.term_months as f64;
    let total_interest = (total_of_payments - i.principal_usd).max(0.0);

    let maturity = NaiveDate::parse_from_str(&i.start_date, "%Y-%m-%d")
        .ok()
        .and_then(|d| d.checked_add_months(Months::new(i.term_months)))
        .map(|d| d.format("%Y-%m-%d").to_string())
        .unwrap_or_default();

    let citation = i.statute_citation.trim();
    let pursuant = if citation.is_empty() {
        format!(
            "This Note shall be governed by the laws of the State of {}.",
            i.governing_state
        )
    } else {
        format!(
            "This Note shall be governed by the laws of the State of {} ({}).",
            i.governing_state, citation
        )
    };

    let mut clauses = vec![
        DocClause {
            heading: "Parties".into(),
            body: format!(
                "Borrower: {}\nLender: {}\nPrincipal Amount: {}\nAnnual Interest Rate: {:.3}%\nTerm: {} months",
                i.borrower_name,
                i.lender_name,
                money(i.principal_usd),
                i.annual_rate_pct,
                i.term_months
            ),
        },
        DocClause {
            heading: "1. Promise to Pay".into(),
            body: format!(
                "For value received, {} (\"Borrower\") promises to pay to the order of {} (\"Lender\") the principal sum of {}, together with interest as set forth below.",
                i.borrower_name, i.lender_name, money(i.principal_usd)
            ),
        },
        DocClause {
            heading: "2. Interest".into(),
            body: format!(
                "Interest shall accrue on the unpaid principal balance at the rate of {:.3}% per annum.",
                i.annual_rate_pct
            ),
        },
        DocClause {
            heading: "3. Repayment".into(),
            body: format!(
                "Borrower shall repay this Note in {} equal monthly installments of {} each, beginning on {}, with a final maturity date of {}, on which date the entire remaining balance, if any, shall be due and payable. Total of payments over the term: {} (of which {} is interest).",
                i.term_months,
                money(payment),
                i.start_date,
                maturity,
                money(total_of_payments),
                money(total_interest)
            ),
        },
        DocClause {
            heading: "4. Prepayment".into(),
            body: "Borrower may prepay this Note in whole or in part at any time without penalty. Each prepayment shall be applied first to accrued interest and then to principal.".into(),
        },
    ];

    let mut next = 5;
    if i.late_fee_usd > 0.0 {
        clauses.push(DocClause {
            heading: format!("{next}. Late Charge"),
            body: format!(
                "If any installment is not paid within ten (10) days of its due date, Borrower shall pay a late charge of {} for that installment.",
                money(i.late_fee_usd)
            ),
        });
        next += 1;
    }

    clauses.push(DocClause {
        heading: format!("{next}. Default and Acceleration"),
        body: "If Borrower fails to pay any installment when due and the default continues, Lender may declare the entire unpaid principal balance and accrued interest immediately due and payable, and may pursue any remedy available at law or in equity.".into(),
    });
    next += 1;

    clauses.push(DocClause {
        heading: format!("{next}. Governing Law"),
        body: pursuant,
    });

    clauses.push(DocClause {
        heading: "Signatures".into(),
        body: format!(
            "Borrower: ____________________  Date: __________\n{}\n\nLender: ____________________  Date: __________\n{}",
            i.borrower_name, i.lender_name
        ),
    });

    PromissoryNote {
        title: "Promissory Note".into(),
        monthly_payment_usd: payment,
        total_of_payments_usd: total_of_payments,
        total_interest_usd: total_interest,
        maturity_date: maturity,
        statutory_citation: citation.to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn base() -> NoteInput {
        NoteInput {
            lender_name: "Jane Lender".into(),
            borrower_name: "John Borrower".into(),
            principal_usd: 10_000.0,
            annual_rate_pct: 6.0,
            term_months: 12,
            start_date: "2026-01-01".into(),
            late_fee_usd: 0.0,
            governing_state: "Texas".into(),
            statute_citation: String::new(),
        }
    }

    #[test]
    fn monthly_payment_amortizes() {
        // $10k, 6%/yr, 12 months → ~$860.66.
        let n = generate(&base());
        assert!((n.monthly_payment_usd - 860.66).abs() < 0.01);
    }

    #[test]
    fn totals_consistent() {
        let n = generate(&base());
        // total ≈ 10327.97, interest ≈ 327.97.
        assert!((n.total_of_payments_usd - 10_327.97).abs() < 0.05);
        assert!((n.total_interest_usd - 327.97).abs() < 0.05);
        // Identity: total = principal + interest.
        assert!((n.total_of_payments_usd - (10_000.0 + n.total_interest_usd)).abs() < 1e-6);
    }

    #[test]
    fn maturity_is_start_plus_term() {
        // 2026-01-01 + 12 months = 2027-01-01.
        assert_eq!(generate(&base()).maturity_date, "2027-01-01");
    }

    #[test]
    fn zero_rate_is_straight_principal_split() {
        let n = generate(&NoteInput {
            principal_usd: 12_000.0,
            annual_rate_pct: 0.0,
            ..base()
        });
        assert!((n.monthly_payment_usd - 1_000.0).abs() < 1e-6);
        assert!((n.total_interest_usd - 0.0).abs() < 1e-6);
    }

    #[test]
    fn late_fee_clause_only_when_set() {
        let without = generate(&base());
        assert!(!without.clauses.iter().any(|c| c.heading.contains("Late Charge")));
        let with = generate(&NoteInput { late_fee_usd: 25.0, ..base() });
        assert!(with.clauses.iter().any(|c| c.heading.contains("Late Charge")));
    }

    #[test]
    fn clause_numbering_stays_sequential_with_late_fee() {
        // With the optional late-charge clause present, Governing Law must still
        // be numbered one past it (no gap, no collision).
        let with = generate(&NoteInput { late_fee_usd: 25.0, ..base() });
        assert!(with.clauses.iter().any(|c| c.heading == "5. Late Charge"));
        assert!(with.clauses.iter().any(|c| c.heading == "6. Default and Acceleration"));
        assert!(with.clauses.iter().any(|c| c.heading == "7. Governing Law"));
    }

    #[test]
    fn statute_citation_echoed() {
        let n = generate(&NoteInput {
            statute_citation: "Tex. Fin. Code § 302".into(),
            ..base()
        });
        assert_eq!(n.statutory_citation, "Tex. Fin. Code § 302");
        assert!(n.clauses.iter().any(|c| c.body.contains("Tex. Fin. Code § 302")));
    }

    #[test]
    fn bad_date_yields_empty_maturity() {
        let n = generate(&NoteInput { start_date: "nope".into(), ..base() });
        assert_eq!(n.maturity_date, "");
    }
}
